Zephyr add possibility to use littlefs with blk MMC devices without flash map

xakep-amatop opened the pull request:
LittleFS may be used not only for Flash devices, we can use it for block
devices too.

So, I have changed dependency rules: add dependency of config on
MMC_STACK. Add possibility of building LittleFS driver for case when
CONFIG_FLASH_MAP and CONFIG_FLASH_PAGE_LAYOUT are disabled.

de-nordic commented:
@xakep-amatop Have you tried the Kconfig option `CONFIG_FS_LITTLEFS_BLK_DEV` ?

Laczen commented:
@de-nordic, I think @xakep-amatop wants to use littlefs without flash support, which IIRC is not possible at the moment. I have created a PR in the past to clean up the littlefs setup to allow for this (#41972), but this was never included.

de-nordic commented:
> @de-nordic, I think @xakep-amatop wants to use littlefs without flash support, which IIRC is not possible at the moment. I have created a PR in the past to clean up the littlefs setup to allow for this (#41972), but this was never included.

Understood. Right the dependency on FLASH stuff comes directly from LITTLEFS.

de-nordic commented on include/zephyr/drivers/flash.h:
```diff
@@ -317,6 +317,15 @@ typedef bool (*flash_page_cb)(const struct flash_pages_info *info, void *data);
 			void *data);
+#else /* CONFIG_FLASH_PAGE_LAYOUT */
+
+/* function stubs */
+typedef bool (*flash_page_cb)(const struct flash_pages_info *info, void *data);
+
+static inline void
+flash_page_foreach(const struct device *dev, flash_page_cb cb, void *data)
+{}
+
```
No stubbing API headers to make other subsystems happy.

xakep-amatop replied:
fixed

de-nordic commented on include/zephyr/storage/flash_map.h:
```diff
@@ -106,6 +106,8 @@ int flash_area_check_int_sha256(const struct flash_area *fa,
 				const struct flash_area_check *fac);
 #endif
 
+#if defined(CONFIG_FLASH_MAP)
```
Oh no. If entire API header has to be conditional, maybe it should not be included in the first place.

xakep-amatop replied:
fixed

de-nordic commented on subsys/fs/Kconfig.littlefs:
```diff
@@ -6,8 +6,7 @@
 	depends on FILE_SYSTEM
-	depends on FLASH_MAP
-	depends on FLASH_PAGE_LAYOUT
+	depends on (FLASH_MAP && FLASH_PAGE_LAYOUT) || MMC_STACK
```
Instead of doing this, the approach to Kconfig from https://github.com/zephyrproject-rtos/zephyr/pull/41972/files#diff-b12838b5f8feaf474fa81aeab169e1a7ac9765d34c81f76845f658425bd71d79R17 should be considered, where there are two separate options to allow FLASH_MAP based and DISK_ACCESS based LittleFS independently, as user may use both of them at runtime.

xakep-amatop replied:
done

de-nordic reviewed (changes requested):
-1 for modification of API headers.

xakep-amatop committed:
fs: littlefs: remove dependency on flash map from FILE_SYSTEM_LITTLEFS

LittleFS may be used not only for Flash devices, we can use it for block
devices too. So, I have changed dependency rules and remove dependency
on flash map from FILE_SYSTEM_LITTLEFS.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>
Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
fs: littlefs: add ability to build little fs without flash map

LittleFS may be used not only for Flash devices, we can use it for
block devices too.

Add possibility of building LittleFS for case when
CONFIG_FLASH_MAP and CONFIG_FLASH_PAGE_LAYOUT are disabled.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
samples: littlefs: remove flash map configs when we use block device

We don't need anymore enabling of Flash map for building littlefs
sample with block disk device config.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
> -1 for modification of API headers.

fixed

de-nordic reviewed (changes requested):
Fine with addition of new Kconfig option to separately enable flash map and blk device operation, but code still requires some changes.
Also, please add commit with release notes update informing on addition of Kconfig option and how does it now impact LittleFS configuration for user.

xakep-amatop commented on subsys/fs/littlefs_fs.c:
```diff
@@ -190,6 +199,24 @@ static int lfs_api_erase(const struct lfs_config *c, lfs_block_t block)
+static int lfs_api_erase(const struct lfs_config *c, lfs_block_t block)
+{
+	return errno_to_lfs(-EIO);
+}
```
done

xakep-amatop committed:
doc: release: 3.4: add note about adding FS_LITTLEFS_FMP_DEV

Update release notes 3.4 with the new Flash Map Kconfig option
for LittleFS.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

# Conflicts:
#	doc/releases/release-notes-3.4.rst

xakep-amatop commented:
note: I've fixed merge conflicts

de-nordic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -646,30 +658,41 @@ static int littlefs_flash_init(struct fs_littlefs *fs, uintptr_t dev_id)
+	__ASSERT_NO_MSG((IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && !littlefs_on_blkdev(flags)) ||
+			(IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && littlefs_on_blkdev(flags)));
+
+#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
 	if (littlefs_on_blkdev(flags)) {
 		fs->backend = (void *) dev_id;
```
I have noticed that all #ifdefs added within source code are around `if`. So we have something like:
```
#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
	if (littlefs_on_blkdev(flags)) {
```
Zephyr provides `IS_ENABLED` macro that evaluates C preprocessor identifier to 0 or 1, which could be used here:
```
if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && littlefs_on_blkdev(flags)) {
```
This makes code look nicer, as it is not so jagged by the #fidef.

xakep-amatop replied:
If I understand you correctly, you mean dead-code elimination (DCE) here.
Compiler parses code before the DCE optimization. It is meant that shouldn't be undefined symbols and code has to be compilable, otherwise the compilation process will fail even if we expect DCE.
If the BLK or FMP configs are disabled, a lot of functions in this file will be undefined, i.e. we need to add stub functions (I've deleted some of them according to yours previous request) or leave the current approach or enable both configs by default, but the last option contradicts the purpose of this pull-request

de-nordic replied:
OK, you are right. I have forgotten about function definitions.

de-nordic commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -486,6 +486,8 @@ Libraries / Subsystems
   * The FAT FS initialization order has been updated to match LittleFS, fixing an issue where
     attempting to mount the disk in a global function caused FAT FS to fail due to not being registered beforehand.
     FAT FS is now initialized in POST_KERNEL.
+  * Added :kconfig:option:`CONFIG_FS_LITTLEFS_FMP_DEV` to enable possibility of using
+    LittleFS for block devices only, e.g. without Flash support.
```
Note that the option is default y too keep previous behaviour.

xakep-amatop replied:
thanks! Added

xakep-amatop commented on subsys/fs/littlefs_fs.c:
```diff
@@ -646,30 +658,41 @@ static int littlefs_flash_init(struct fs_littlefs *fs, uintptr_t dev_id)
 {
+#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
 	int ret = 0;
+#endif
```
fixed

de-nordic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -646,31 +658,38 @@ static int littlefs_flash_init(struct fs_littlefs *fs, uintptr_t dev_id)
 {
 	int ret = 0;
 
+	__ASSERT_NO_MSG((IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && !littlefs_on_blkdev(flags)) ||
```
No. If somebody makes mistake the device can get in  brick loop, when attempting to mount system.
I know that this should be catch in tests, but if tests fail to discover the probelI know that this is impossible situation, or will happen once and gets fixed, but I would rather have -ENOTSUP reported in such situation.

Better make it so that if flags are not correctly selected the init returns with error but allows device to work, so that at least it could be updated or have logs pulled remotely,

de-nordic replied:
I know that there are __ASSERTS used in other parts of LittleFS code, but I would avoid it as much as possible.

xakep-amatop replied:
done

de-nordic reviewed (changes requested):
One more change requested, otherwise looks ok.
